src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -784,11 +779,18 @@ internal ReadResult GetReadAsyncResult()
                 ThrowHelper.ThrowInvalidOperationException_GetResultNotCompleted();
             }
 
+            ReadResult result;
+            CancellationTokenRegistration cancellationTokenRegistration;
+
             lock (_sync)
             {
-                GetReadResult(out ReadResult result);
-                return result;
+                GetReadResult(out result);
+
+                cancellationTokenRegistration = _readerAwaitable.ReleaseCancellationTokenRegistration();
             }
+
+            cancellationTokenRegistration.Dispose();
```

Comment:
Is this guaranteed to be called even if the the PipeWriter is completed with an error during PipeReader.ReadAsync()?